src/Microsoft.AspNetCore.SignalR.Common/Internal/Protocol/Utf8BufferTextWriter.cs

```diff
@@ -86,36 +87,43 @@ public override void Write(char[] buffer)
 
         public override void Write(char value)
         {
-            var destination = GetBuffer();
-
             if (value <= 127)
             {
-                destination[0] = (byte)value;
+                EnsureBuffer();
+
+                // Only need to set one byte
+                // Avoid Memory<T>.Slice overhead for perf
+                _memory.Span[_memoryUsed] = (byte)value;
                 _memoryUsed++;
             }
             else
             {
-                // Json.NET only writes ASCII characters by themselves, e.g. {}[], etc
-                // this should be an exceptional case
-                var bytesUsed = 0;
-                var charsUsed = 0;
-                unsafe
-                {
+                WriteMultiByteChar(value);
+            }
+        }
+
+        private unsafe void WriteMultiByteChar(char value)
+        {
+            var destination = GetBuffer();
+
+            // Json.NET only writes ASCII characters by themselves, e.g. {}[], etc
+            // this should be an exceptional case
+            var bytesUsed = 0;
+            var charsUsed = 0;
 #if NETCOREAPP2_1
-                    _encoder.Convert(new Span<char>(&value, 1), destination, false, out charsUsed, out bytesUsed, out _);
+            _encoder.Convert(new Span<char>(&value, 1), destination, false, out charsUsed, out bytesUsed, out _);
 #else
-                    fixed (byte* destinationBytes = &MemoryMarshal.GetReference(destination))
-                    {
-                        _encoder.Convert(&value, 1, destinationBytes, destination.Length, false, out charsUsed, out bytesUsed, out _);
-                    }
+            fixed (byte* destinationBytes = &MemoryMarshal.GetReference(destination))
+            {
+                _encoder.Convert(&value, 1, destinationBytes, destination.Length, false, out charsUsed, out bytesUsed, out _);
+            }
 #endif
-                }
-                Debug.Assert(charsUsed == 1);
 
-                if (bytesUsed > 0)
-                {
-                    _memoryUsed += bytesUsed;
-                }
+            Debug.Assert(charsUsed == 1);
+
+            if (bytesUsed > 0)
```

Comment:
Why bother checking bytesUsed? Just add to _memoryUsed.

Reply:
ah, because this code use to call Advance, and Advance(0) would a-splode